### DISCUSSION TITLE
TUI: Features and design

### DISCUSSION THREAD

**Discussion by TornaxO7** [pimalaya/himalaya#332] [Ideas]
I'd like to collect the ideas and stuff about the TUI in this discussion. Feel free to shoutout any idea/suggestion for the TUI and what your expectations are.

**Comment by TornaxO7** [comment]
@soywod is it possible to pin this somewhere?

**Reply by soywod** [reply]
Done :wink: nice idea btw!

**Comment by yara-blue** [comment]
I use neomutt for my mail right now but I would love to switch to (and maybe work on?) a himalaya based backend. Here are some (rough) ideas of what I would like to see.

- It woud be nice if the TUI had a [sidebar](https://neomutt.org/feature/sidebar-intro) like _neomutt_. The items in the side bar could be separate mailboxes/accounts. They could also be the result of user defined scripts, like email folders/labels. All done locally. 
- Urls should be clickable even if they span across multiple lines. Urls could even be shortend by default and expanded on keypress? This way they do not take up the entire email if they are extremely long (rather common).
- A config file, could be in `.toml` or `.lua` like _neovim_. Using lua opens up the ability to (in the future) add custom behavior. For example defining lua functions that sort mails into folder. Using lua a user can, instead of a string, have a lua command in the config to set the password (maybe calling a password manager).
- Do not support editing text, this can be left to another program such as: _vim_, _emacs_ or _nano_. The default can be picked from the `VISUAL` or `EDITOR` env var. The config should expose a way to override the default. This would allow me to set up a separate _neovim_ config for email.
- Fuzzy email search to text (like [telescope](https://github.com/nvim-telescope/telescope.nvim)).
- Smart defaults that adapt to the current context. If I am in mailbox _A_ I want my mail to be signed with _sig_a_ and send from _adress_a_ however in mailbox _B_ these must be different still. For mailboxes _C_ and _D_ I have configured nothing and want the default.

**Reply by soywod** [reply]
> * A config file, could be in `.toml` or `.lua` like _neovim_. Using lua opens up the ability to (in the future) add custom behavior. For example defining lua functions that sort mails into folder. Using lua a user can, instead of a string, have a lua command in the config to set the password (maybe calling a password manager).

I like the idea of using a runtime language, but I would limit it to the minimum (for config only). I found this interesting project that could help us interacting with Lua: https://github.com/amethyst/rlua

> * Do not support editing text, this can be left to another program such as: _vim_, _emacs_ or _nano_. The default can be picked from the `VISUAL` or `EDITOR` env var. The config should expose a way to override the default. This would allow me to set up a separate _neovim_ config for email.

I forgot where I saw this (maybe aerc?), but it is possible to embed the editor inside the TUI (which avoid the fact to get out of the TUI), sth to dig.

> * Fuzzy email search to text (like [telescope](https://github.com/nvim-telescope/telescope.nvim)).

This sounds difficult. Plus it is competing directly with Notmuch. Sth to dig also!

> * Smart defaults that adapt to the current context. If I am in mailbox _A_ I want my mail to be signed with _sig_a_ and send from _adress_a_ however in mailbox _B_ these must be different still. For mailboxes _C_ and _D_ I have configured nothing and want the default.

I think you are referring to Himalaya's account: they act like contexts. So it is already the case. What we need is the ability to change account automatically when switching mailboxes.

For the rest, I agree. Thank you for the ideas :pray:

**Reply by yara-blue** [reply]
> > * Do not support editing text, this can be left to another program ....
> 
> I forgot where I saw this (maybe aerc?), but it is possible to embed the editor inside the TUI (which avoid the fact to get out of the TUI), sth to dig.

That is a great idea, aerc seems to do this by embedding a text terminal inside its TUI in which the editor runs. (Neo)mutt must be doing something similarly. There is the danger of a user getting stuck. In vim happens when opening a vim instance inside a vim terminal (happens when you do git commit and git starts vim...). 

> > * Fuzzy email search to text (like [telescope](https://github.com/nvim-telescope/telescope.nvim)).
> 
> This sounds difficult. Plus it is competing directly with Notmuch. Sth to dig also!

I did not think about mail indexing and I have not tried notmuch. In my opinion however good mail search should, at one point, be a part of the TUI. Setting up (neo)mutt with all the tooling around it is quite the chore. Having installed notmuch you will still have to set up another tool to download and keep mail up to date. On (ubuntu) linux this is doable but I have no clue how this is done on mac, or god forbid windows, I would guess a lot of compiling is involved.

This leads to the larger question: build a minimal TUI relying on existing proven programs or build an extendable monolith reinventing the wheel for ease of use and setup?

Without indexing fuzzy search could use this existing [library](https://crates.io/crates/fuzzy-matcher). For indexed search [tantivy](https://github.com/quickwit-oss/tantivy) could be integrated. Both would require storing mail locally or downloading mail during the search.

> > * Smart defaults that adapt to the current context...
> 
> I think you are referring to Himalaya's account: they act like contexts. So it is already the case. What we need is the ability to change account automatically when switching mailboxes.

Multiple mailboxes/labels/folders could be under the same smpt account. I have a mailbox.org account where have multiple mail aliases on my domain. Mail can be send to me@domain.ext or git@domain.ext they all arrive on the same smpt server/account (mailbox.org). I want mail from the different users, here _me_ and _git_, to go into separate mailboxes/labels/folders. If I then reply from a folder I want the TUI to pick the right _from_ adress.

**Reply by soywod** [reply]
> That is a great idea, aerc seems to do this by embedding a text terminal inside its TUI in which the editor runs. (Neo)mutt must be doing something similarly. There is the danger of a user getting stuck. In vim happens when opening a vim instance inside a vim terminal (happens when you do git commit and git starts vim...).

Yes, it needs to be well done!

> I did not think about mail indexing and I have not tried notmuch. In my opinion however good mail search should, at one point, be a part of the TUI. Setting up (neo)mutt with all the tooling around it is quite the chore. Having installed notmuch you will still have to set up another tool to download and keep mail up to date. On (ubuntu) linux this is doable but I have no clue how this is done on mac, or god forbid windows, I would guess a lot of compiling is involved.

Totally agree on the fact that the TUI should be able to search. As I also said in [another post](https://github.com/soywod/himalaya/pull/155#issuecomment-1060363876), it should also work offline (they are both linked). I had in mind to synchronize (in background, it should be transparent for the user, think Thunderbird) the IMAP backend with a local Maildir. Then we can search easily inside (tantivy sounds a good choice).

> This leads to the larger question: build a minimal TUI relying on existing proven programs or build an extendable monolith reinventing the wheel for ease of use and setup?

Good question. I would say sth in the middle. Let me explain.

The problem with all these mainstream tools is that:
- they are ALL TUIs, which is really restrictive, and which also means you need to configure them a lot to fit more or less your workflow
- they reinvent the wheel: fetching an email from an IMAP server, list emails from a Maildir etc

With Himalaya I try to fix those issues by separating the logic from the UI. This way you have more chance to find the right UI to fit your workflow (Vim plugin, Emacs plugin, CLI, TUI, GUI, Web…), and each UI does not need to reinvent the wheel.

In the long term, I see Himalaya as a simple Rust library containing the logic to connect to all kind of backends, and an army of frontends connecting to it. Aerc or Neomutt could even use it and became simple UIs!

> Without indexing fuzzy search could use this existing [library](https://crates.io/crates/fuzzy-matcher). For indexed search [tantivy](https://github.com/quickwit-oss/tantivy) could be integrated. Both would require storing mail locally or downloading mail during the search.

This will be doable with the feature I mentioned earlier (sync IMAP with a local Maildir).

**Reply by TornaxO7** [reply]
> but it is possible to embed the editor inside the TUI (which avoid the fact to get out of the TUI), sth to dig.

Hm... what do you mean with "embed"? You mean that the $EDITOR should be called, right?

**Reply by soywod** [reply]
> Hm... what do you mean with "embed"? You mean that the $EDITOR should be called, right?

Yes, to use `$EDITOR` and to display the editor inside the TUI (see this as a web iframe).

**Reply by yara-blue** [reply]
> In the long term, I see Himalaya as a simple Rust library containing the logic to connect to all kind of backends, and an army of frontends connecting to it. Aerc or Neomutt could even use it and became simple UIs!

In a few months when I have some time to throw at it I would love to supply the army of frontends. Thanks a lot for your time and work!

**Reply by TornaxO7** [reply]
> Yes, to use $EDITOR and to display the editor inside the TUI (see this as a web iframe).

I disagree here. I'd expect a terminal application to be able to run the in TTY as well. So running a web iframe won't be possible in the tty.

**Reply by soywod** [reply]
This can be set by an option? So the user has the choice to display the editor inside the TUI or not?

**Reply by TornaxO7** [reply]
> This can be set by an option? So the user has the choice to display the editor inside the TUI or not?

Yes, that would be possible. But I don't see the benefit if we put the window of `$EDITOR` into the TUI. Why should we not just call `$EDITOR` as a shell-command?

**Reply by soywod** [reply]
Not to get out of the TUI. It is a bit like on the web: I prefer to have my content inside my webpage in an iframe instead of another tab or window.

**Reply by yara-blue** [reply]
It would be quite nice to have new drafts be typed in a split. The upper half of the screen would be some already open email and the lower half a `tty` with `$EDITOR` running

**Reply by TornaxO7** [reply]
> Not to get out of the TUI. It is a bit like on the web: I prefer to have my content inside my webpage in an iframe instead of another tab or window.

But this is going to be tricky with the keybindings. How should himalaya know, if your keys, which you're typing are refering to himalaya and not to the editor?

**Reply by yara-blue** [reply]
> > Not to get out of the TUI. It is a bit like on the web: I prefer to have my content inside my webpage in an iframe instead of another tab or window.
> 
> But this is going to be tricky with the keybindings. How should himalaya know, if your keys, which you're typing are refering to himalaya and not to the editor?

Simplest way is simply to disable the use of the rest of the TUI until the editor is closed. Alternatively you could set up a keybind to escape the editor window (then keys will not be captured by the editor). It needs some thought but the problem has been solved in other TUI with a tty.

**Reply by TornaxO7** [reply]
> Simplest way is simply to disable the use of the rest of the TUI until the editor is closed.

Well, but what is, if you need to read another mail which includes some information which you need for your answer? Should the user open a new himalaya instance?

> Alternatively you could set up a keybind to escape the editor window (then keys will not be captured by the editor). It needs some thought but the problem has been solved in other TUI with a tty

I don't know about this... this could be very tricky. You'd have to find another key-binding just for switching between the editor and himalaya. Some people might have only "bad" keybindings left which can be mapped.

**Reply by soywod** [reply]
Anyway, this remains an idea, if it ends up to be too difficult to implement we can just skip it :wink:

**Comment by TornaxO7** [comment]
About the UI of the TUI:

I was thinking about the following main features for the UI:
- a sidebar (for the different mailboxes)
- tabs (for switching accounts)
- main-list panel (where the mails are listed)
- (optional) a little preview-panel on the right which shows the content of the mail.

But I also though that it might be possible to let the user customize what panel should include what content or even how the layout should be. Some people might prefer to have the different accounts in the sidebar on the left and the mailboxes in the tabs for example.

## TL;DR

I could imagine the following layout:
```
| tab 1| tab 2 | tab 3 |
----------------------------------------------------
|     First panel     | Second panel | Third panel |
|                     |              |             |
|                     |              |             |
|                     |              |             |
|                     |              |             |
---------------------------------------------------|
```
And the user should be able to set the ratio of the panels and also (probably) their position. For example one panel can be even at the bottom. The user can also decide which panel should include which content.
I don't know if you can actually change the position of the tabs to the side as well or at the bottom, but I can look if it's possible with this.

## When reading mails
If you open up a mail, the following UI should be loaded:

```
| tab 1 | tab 2 | tab 3|
-----------------------|-----------------|
| From:                | - Attachment 1  |
| Subject:             | - Attachment 2  |
| CC:                  | - Attachment 3  |
| Date:                |                 |
-----------------------|-----------------|
| Content of the mail                    |
|                                        |
|                                        |
|                                        |
|                                        |
|                                        |
|----------------------------------------|
```

Hm...I'm a little bit unsure how we should handle hyperlinks... also heavy html
docs are only partially able to display them in the terminal. I'd suggest here
to open a browser for this which loads the html-mail.

There should also be an option to display *all* headers in the mail.

## Writing a mail
When writing a new mail, the following should be displayed:

```
| tab 1 | tab 2 | tab 3 |
|----------------------|-Attachments----|
| To:                  |                |
| From:                |                |
| Date: <auto-inserted>|                |
| Subject:             |                |
| CC:                  |                |
| BCC:                 |                |
|----------------------|----------------|
| Content of the mail                   |
|                                       |
|                                       |
|                                       |
|                                       |
|---------------------------------------|
```

**Reply by yara-blue** [reply]
> Hm...I'm a little bit unsure how we should handle hyperlinks

I would suggest to insert a number next to the hyperlink. On pressing the number the default browser would open that link. The number should have a different color so it can easily be recognized as not part of the mail.

> I'd suggest here to open a browser for this which loads the html-mail.

I personally hate html emails. Usually it is enough to recognize the text fields and only show them. Right now I use `w3m -I %{charset} -T text/html` to show html mails from neomutt.

**Reply by TornaxO7** [reply]
> I would suggest to insert a number next to the hyperlink. On pressing the number the default browser would open that link. The number should have a different color so it can easily be recognized as not part of the mail.

Hm... yes, that could be a solution.

> I personally hate html emails. Usually it is enough to recognize the text fields and only show them. Right now I use w3m -I %{charset} -T text/html to show html mails from neomutt.

The problem here is, that `w3m` just displays them as a text (obviously) and some html mails rely on the correct formatting. At least I had a bad experience with `w3m` + mails as text...

**Reply by yara-blue** [reply]
Mmm there is pandoc for converting html to markdown. Then we could show the markdown. Alternatively a library could be written to convert html emails to text. Can you expand on the formatting issues with w3m converted text?

**Reply by TornaxO7** [reply]
> Can you expand on the formatting issues with w3m converted text?

Phew... not really, because I haven't used `w3m` for a long time now... I think I saw stuff like `<tag>dcuedum8e028824539<stuff>` in my mail which made it impossible for me to read the content of the mail. After opening it up in the browser, I could finally see the content clearly.

**Reply by yara-blue** [reply]
Probably best to put this in a config, if lua is used as config users can call w3m, open in a browser, call something else or run type up their own lua code. Default behavior could be open in browser.

**Comment by soywod** [comment]
Here are my plans for the `v0.6.0`. It is not *directly* related to the TUI but let's say it prepares everything around to welcome the TUI (and future frontends) within the best conditions. Please feel free to challenge it https://github.com/soywod/himalaya/discussions/345.
